|                |   |
:----------------|---|
| **Nombre**     |Santiago Escutia Ríos   |
| **Fecha**      | 23/4/2026  |
| **Expediente** |757839   |

In [6]:
import pandas as pd
df= pd.read_excel("Alturas.xlsx")

In [7]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 16 entries, 0 to 15
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Registro      16 non-null     int64  
 1   Sexo          16 non-null     str    
 2   Edad          16 non-null     int64  
 3   Estatura (m)  16 non-null     float64
dtypes: float64(1), int64(2), str(1)
memory usage: 644.0 bytes


In [8]:
df.head()

,Registro,Sexo,Edad,Estatura (m)
0,1,H,21,1.73
1,2,M,19,1.62
2,3,H,21,1.73
3,4,H,19,1.69
4,5,H,20,1.75


In [9]:
# Promedio General
promedio_general = df['Estatura (m)'].mean()

# Promedio para Hombres (H)
promedio_hombres = df[df['Sexo'] == 'H']['Estatura (m)'].mean()

# Promedio para Mujeres (M)
promedio_mujeres = df[df['Sexo'] == 'M']['Estatura (m)'].mean()

print(f"Promedio General: {promedio_general:.2f}")
print(f"Promedio Hombres (H): {promedio_hombres:.2f}")
print(f"Promedio Mujeres (M): {promedio_mujeres:.2f}")

Promedio General: 1.71
Promedio Hombres (H): 1.72
Promedio Mujeres (M): 1.63


# A07 - Árboles de regresión

En esta etapa se cargan la librerías y leemos el archivo de publicidad para definir nuestras variables. Usaremos la inversión en TV como predictor para las Ventas (Sales).

In [14]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

df = pd.read_csv('Advertising.csv')
df.head()

,Unnamed: 0,TV,radio,newspaper,sales
0,1,230.1,37.8,69.2,22.1
1,2,44.5,39.3,45.1,10.4
2,3,17.2,45.9,69.3,9.3
3,4,151.5,41.3,58.5,18.5
4,5,180.8,10.8,58.4,12.9


In [15]:
X = df[["TV", "radio", "newspaper"]]
y = df['sales']

Comparamos qué tan bien se adaptan una regresión lineal y un árbol de decisión usando el 100% de la información. Esto nos sirve para ver la capacidad de ajuste inicial de cada modelo.

In [16]:
#Regresión Lineal Simple
modelo_lr = LinearRegression()
modelo_lr.fit(X, y)
r2_lineal_total = r2_score(y, modelo_lr.predict(X))

# Arbol de Decisión
modelo_tree = DecisionTreeRegressor(random_state=42)
modelo_tree.fit(X, y)
r2_tree_total = r2_score(y, modelo_tree.predict(X))

print(f"R2 Lineal (Total): {r2_lineal_total:.4f}")
print(f"R2 Árbol (Total):  {r2_tree_total:.4f}")

R2 Lineal (Total): 0.8972
R2 Árbol (Total):  1.0000


Para probar si los modelos realmente aprenden o solo memorizan, dividimos los datos a la mitad. Entrenamos con una parte y evaluamos con la otra para medir su precisión real con datos desconocidos.

In [17]:
# 50/50 Train-Split test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=42)

In [ ]:
# Regresión Lineal con el 50/50
modelo_lr_split = LinearRegression().fit(X_train, y_train)
y_pred_lr = modelo_lr_split.predict(X_test)
r2_lineal_split = r2_score(y_test, y_pred_lr)
r2_li

# Árbol con el 50/50
modelo_tree_split = DecisionTreeRegressor(random_state=42).fit(X_train, y_train)
y_pred_tree = modelo_tree_split.predict(X_test)
r2_tree_split = r2_score(y_test, y_pred_tree)


print(f"R2 Lineal (Split 50/50): {r2_lineal_split:.4f}")
print(f"R2 Árbol (Split 50/50):  {r2_tree_split:.4f}")

R2 Lineal (Split 50/50): 0.8721
R2 Árbol (Split 50/50):  0.9345


In [28]:
lr = LinearRegression()

lr.fit(X_train, y_train)

r2_train_tree = lr.score(X_train, y_train)
r2_test_tree = lr.score(X_test, y_test)

print("R2 en train para lineal:", r2_train_tree)
print("R2 en test para lineal:", r2_test_tree)

R2 en train para lineal: 0.9042613648908893
R2 en test para lineal: 0.8721004816045135


In [29]:
tree = DecisionTreeRegressor(random_state=42)


tree.fit(X_train, y_train)
r2_train_tree = tree.score(X_train, y_train)
r2_test_tree = tree.score(X_test, y_test)

print("R2 en train para tree:", r2_train_tree)
print("R2 en test para tree:", r2_test_tree)

R2 en train para tree: 1.0
R2 en test para tree: 0.9344975035059896


El árbol de decisión sacó un 1.0 perfecto en entrenamiento porque memorizó los datos (sobreajuste). Aunque su resultado en la prueba es bueno (0.93), la regresión lineal es más equilibrada y realista. Por lo que el modelo lineal es preferible porque es más estable y no se engaña con datos aprendidos de memoria.